perf(core): build-in for should update indexes only when a change is detected

pkozlowski-opensource opened the pull request:
This change adjust the built-in for algorithm of dealing with embedded views to update the repeater context (and more specifically - its index field) only when a change between collections was detected.

alxhub reviewed (approved):
Overall this LGTM as far as the logic goes.

`reconcile` returning a boolean for whether or not index updates are needed though is a little bit awkward, since it's somewhat arbitrary. Normally I would expect a boolean return value to indicate success or failure or something.

Alternatively, could we hang the index updating logic on the `LiveCollection` instead? Then `reconcile` could call `liveCollection.updateIndices()` if it detects differences, rather than returning a boolean to request that the caller updates the indices. I think this is in line with the `LiveCollection` concept/design.

pkozlowski-opensource committed:
perf(core): build-in for should update indexes only when views were added / removed

This change adjust the built-in for algorithm of dealing with embedded views
to update the repeater context (and more specifically - its index field)
only when views were added / removed "in the middle of LContainer" (in places
other than the LContainer end).

This skip iteration over the entire LContainer in most cases - and most importantly
in cases where no diff change was detected.
"

alfaproject commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -154,6 +154,12 @@ export function ɵɵrepeaterCreate(
+   Property indicating if indexes in the repeater context need to be updated following the live
+   collection changes. Index updates are necessary if and only if views are inserted / removed in
+   the middle of LContainer. Adds and removals at the end don't require index updates.
+ */
```
nit: this looks a bit awkward and maybe the reason why spacing is off? Normally this should be:
```js
/**
 * ...
 */
```
